### PR TITLE
Make compliant with draft v16

### PR DESCRIPTION
**Pull request by armfazh**
Make compliant with [draft v16](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-opaque-16)

Merge #18 first

### CODE REVIEW

File: src/3dh_server.ts [hunk 1]
```diff
@@ -27,50 +28,51 @@ export class AKE3DHServer {
...
+    ): Promise<AuthResponse> {
+        const server_nonce = Uint8Array.from(this.config.prng.random(this.config.constants.Nn))
+        const server_keyshare_seed = Uint8Array.from(
+            this.config.prng.random(this.config.constants.Nseed)
         )
-        const ke2 = new KE2(credential_response, auth_response)
+        const { private_key: server_secret, public_key: server_public_keyshare } =
+            await this.config.ake.deriveDHKeyPair(server_keyshare_seed)
         const preamble = preambleBuild(
+            client_identity ? client_identity : client_public_key,
```
Comment (thibmeu) [inline src/3dh_server.ts:39]:
```suggestion
            client_identity ?? client_public_key,
```

File: src/common.ts [hunk 2]
```diff
@@ -226,26 +226,26 @@ export class AKE3DH implements AKEFn {
...
+    //     return { private_key, public_key }
+    // }
+
+    // async generateAuthKeyPair(): Promise<AKEExportKeyPair> {
+    //     const keypair = this.recoverPublicKey(await randomPrivateKey(this.suiteID))
+    //     return {
+    //         private_key: Array.from(keypair.private_key),
+    //         public_key: Array.from(keypair.public_key)
+    //     }
+    // }
```
Comment (thibmeu) [inline src/common.ts:250]: do you mean to remove this, or leave it commented out?
if it's meant to be commented out,. add a comment as to why

File: src/3dh_client.ts [hunk 3]
```diff
@@ -43,16 +45,24 @@ export class AKE3DHClient {
           }
         | Error
     > {
-        if (typeof this.client_secret === 'undefined') {
+        if (typeof this.client_secret === 'undefined' || typeof this.ke1 === 'undefined') {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/3dh_client.ts:48]: done

File: src/common.ts (hunk 2 above)
Comment (armfazh) [reply src/common.ts:250]: ups

File: test/vectors.test.ts [hunk 4]
```diff
@@ -111,37 +109,27 @@ async function createMocks(vector: Vector, cfg: Config, isFake: boolean) {
...
-            public_key: Array.from(notNullHex(vector.inputs.client_keyshare))
-        })
-        .mockResolvedValueOnce({
-            private_key: Array.from(notNullHex(vector.inputs.server_private_keyshare)),
-            public_key: Array.from(notNullHex(vector.inputs.server_keyshare))
+        jest.spyOn(OPRFClient.prototype, 'randomBlinder').mockImplementationOnce(() => {
+            const blind = notNullHex(vector.inputs.blind_login)
+            const group = Oprf.getGroup(cfg.oprf.id as SuiteID)
+            const scalar = group.desScalar(blind)
+            return Promise.resolve(scalar)
```
[the remark this answers is not in the document]
Comment (armfazh) [reply test/vectors.test.ts:117]: good one, updated now

File: test/vectors.test.ts [hunk 5]
```diff
@@ -286,16 +277,87 @@ async function test_full_registration(input: inputTest, vector: Vector): Promise
...
+    // Client
+    // [TODO] We must check that the client tries to complete the
+    // protocol and fails.
+    //
+    // const deser_ke2 = KE2.deserialize(cfg, Array.from(ser_ke2))
+    // const finClient = await client.authFinish(deser_ke2, server_identity, client_identity, context)
+
+    // if (finClient instanceof Error) {
+    //     expect(finClient.message).toBe('EnvelopeRecoveryError')
+    // }
```
[the remark this answers is not in the document]
Comment (armfazh) [reply test/vectors.test.ts:346]: I just worked the extra mile and now this test is already covered.

File: test/vectors.test.ts [hunk 6]
```diff
@@ -334,18 +396,17 @@ async function test_full_login(input: inputTest, vector: Vector): Promise<boolea
...
         deser_ke1,
         credential_file.record,
         credential_file.credential_identifier,
         credential_file.client_identity,
         context
     )
-    expect(ret_auth_init).not.toBeInstanceOf(Error)
-    if (ret_auth_init instanceof Error) {
+    expect(ke2).not.toBeInstanceOf(Error)
+    if (ke2 instanceof Error) {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply test/vectors.test.ts:407]: it's looks redundant, but the first is a test unit, the second is a typescript assertion

[the remark this answers is not in the document]
Comment (thibmeu) [reply test/vectors.test.ts:430]: but the unit test is going to fail if `ke2 instanceof Error` is true.
instead, you could cast the type with `ke2 as <T>`

**Review by thibmeu — APPROVED** [review]
Looks good. no blockers

File: test/vectors.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (armfazh) [reply test/vectors.test.ts]: I wrote type assertion functions `checkNoError` and `checkToBeError` and code was simplified

**Pull request merged**